lib/pbench/server/api/resources/query_apis/query_result.py

```diff
@@ -0,0 +1,257 @@
+        to search.
+
+        The Elasticsearch URL string is also dependent upon the configured
+        "prefix". This is a Pbench artifact that allows multiple Pbench
+        instances to share a single Elasticsearch server, by segregating
+        their data into unique namespaces. Each Pbench index name is
+        qualified by the index_prefix of the Pbench server that originally
+        indexed the data. The prefix comes from the pbench-server.cfg
+        file for the server instance. For example, production, "staging",
+        and development server instances might all share one Elasticsearch
+        cluster by choosing unique prefixes to generate unique index
+        names.
+
+        Required headers include
+
+        X-auth-token:   Pbench session authorization token authorized to access
+                        "user"'s data. E.g., "user", "admin" or a user with
+                        whom the dataset has been shared.
+        Content-Type:   application/json
+        Accept:         application/json
+
+        Return payload is a summary of the "aggregations" property of the
+        returned Elasticsearch query results, showing fields from the "run"
+        subdocument, including the controller and name, the start and end
+        time, and a starting UNIX timestamp.
+
+        TODO: We probably need to return all *unowned* runs when called
+        without a session token or user parameter. We need to define precisely
+        how this should work.
+
+        NOTE: This is the format currently constructed by the Pbench
+        dashboard `src/model/dashboard.js` fetchResults method, which
+        becomes part of the Redux state. (Note that the "key" of the sequence
+        is the same as the "run.name".)
+
+        [
+            {
+                "key": "fio_rhel8_kvm_perf43_preallocfull_nvme_run4_iothread_isolcpus_2020.04.29T12.49.13",
+                "startUnixTimestamp": 1588178953561,
+                "run.name": "fio_rhel8_kvm_perf43_preallocfull_nvme_run4_iothread_isolcpus_2020.04.29T12.49.13",
+                "run.controller": "dhcp31-187.perf.lab.eng.bos.redhat.com",
+                "run.start": "2020-04-29T12:49:13.560620",
+                "run.end": "2020-04-29T13:30:04.918704"
+            }
+        ]
+        """
+        json_data = request.get_json(silent=True)
+        if not json_data:
+            self.logger.info("Invalid JSON object. Query: {}", request.url)
+            abort(400, message="Missing request payload")
+
+        try:
+            user = json_data["user"]
+            name = json_data["name"]
+            start_arg = json_data["start"]
+            end_arg = json_data["end"]
+        except KeyError:
+            keys = [k for k in ("user", "name", "start", "end") if k not in json_data]
+            self.logger.info("Missing required JSON keys {}", ",".join(keys))
+            abort(400, message=f"Missing request data: {','.join(keys)}")
```

Comment:
I've gone back and forth between proactive and optimistic error handling; Pete's strong preference for optimistic code and "wasting time" for diagnosis in post-error handling has definitely affected my style recently.

Of course I could simplify by just reporting the actual `KeyError` and not worrying about any other missing keys, but this seemed more useful since I hate reporting errors one at a time and pushing the client to iterate. (Another "optimistic but inclusive" strategy would be to use separate `try` blocks but join the results before reporting; but that just stretches the code out and makes it more awkward to read for an edge error case, which didn't seem justified.)

Reply:
I certainly appreciate deferring the error-handling and diagnostic code until you need it (the phrase "fast path" leaps to mind for some reason... ;-) ), and I support reporting _all_ the errors instead of making the user come back to find the next one, but, knowing the way that code changes over time, I think it should be reporting the error that it actually found, here.  Now, if you want to report _additional_ errors, that's fine, but the assumption that, because we caught a `KeyError`, it must be one of the keys in `json_data` seems a bit of a stretch, especially as the code evolves in the future.